### PR TITLE
Fix shares read permissions

### PR DESCRIPTION
**Pull request by skjnldsv**
We need to check if getShare (singular) also have resharing rights :)

### CODE REVIEW

**Comment by skjnldsv** [comment]
Wrong turn, we need to use this sanity check!
But it needs to check for resharing rights with `shareProviderResharingRights` :)

**Review by ArtificialOwl — APPROVED** [review]

**Comment by daita** [comment]
So, with the request to displays the list of sharing to users with resharing rights, I missed an endpoint resulting in users not being able to access information they should be allowed to read.

This PR should allow users with resharing rights to retrieve data from a specific `shareId` using [`getShare()`](https://github.com/nextcloud/server/blob/master/apps/files_sharing/lib/Controller/ShareAPIController.php#L300-L317)

nice catch, @skjnldsv

**Review by danxuliu — CHANGES_REQUESTED** [review]
[CI is not happy](https://drone.nextcloud.com/nextcloud/server/20946/41/3); it seems that this gives access to a share to users that should not be able to access it.

Similarly several [sharing integration tests in Talk](https://github.com/nextcloud/spreed/tree/3f274dfdb935045ea07d924718d6f78ed733cd3f/tests/integration/features/sharing) fail too with this pull request (which of course could just mean that they need to be adjusted to the new behaviour, but it seems that they are caused by the same aforementioned problem).

**Comment by skjnldsv** [comment]
Thanks @danxuliu. This is the correct behaviour, so tests should be fixed :)

**Comment by danxuliu** [comment]
The correct behaviour is that a user that has not received a share can get its data and even delete it? That sounds... strange :-P What am I missing? :-)

I would understand it if an ancestor of the share is also shared with that user, but as far as I know in the integration tests the root folder of each user is not shared with other users (and I have just checked it in a development instance and it is possible to delete a share by a user that did not receive it, so it is not only something that happens in the tests).

**Comment by skjnldsv** [comment]
@danxuliu `canAccessShare` means can you access a share, not can you edit it :)
We currently allow a user to see shares of other uses on a file if he have resharing rights to it.

Therefore we need to allow a user that can get the share 123 from the getShares function to also get it from the getShare method.

Though you're right, this is indeed an error! :)
My first throw at this was renaming the canAccessShare to canEditShare. (cc @daita)  But it also conflicts with tons of other permission issues here as well. Like https://github.com/nextcloud/server/blob/2194a02d99e8c86c5d718388beaed46bbade6827/apps/files_sharing/lib/Controller/ShareAPIController.php#L341

File: apps/files_sharing/lib/Controller/ShareAPIController.php [hunk 1]
```diff
@@ -824,12 +830,8 @@ public function updateShare(
 
 		$this->lock($share->getNode());
 
-		if (!$this->canAccessShare($share, false)) {
-			throw new OCSNotFoundException($this->l->t('Wrong share ID, share doesn\'t exist'));
-		}
-
-		if ($share->getShareOwner() !== $this->currentUser && $share->getSharedBy() !== $this->currentUser) {
```
Comment (skjnldsv) [inline apps/files_sharing/lib/Controller/ShareAPIController.php]: Already part of the `canEditShare`

**Comment by skjnldsv** [comment]
All good now!
Please review!

**Comment by skjnldsv** [comment]
All good! Tests added to cover the missing use case!

**Comment by danxuliu** [comment]
Every time you squash unrelated changes to the same commit $DEITY kills a kitten. Please, think of kittens.

I have extracted moving the share features to its own directory and adding the _run-docker.sh_ script to their own commits (I would have added them in separate pull requests instead of this one, but that is a different matter :-P), fixed a typo in the .drone.yml file (_sharing-features/_ was used instead of _sharing_features/_), reordered the pipelines to match the movement of files (`git diff --histogram` shows it better than the default diff algorithm), and adjusted the script documentation to the server.

Regarding the fixes themselves the integration tests in the server pass, although some in Talk fail. One of those tests fail because, before, when you tried to update a received share (that you had access to) but the permissions did not allow you to do that you got a 403, but now you get a 404. I think that the previous value was the right one, as the share is found but you are forbidden to do the operation. However, I tried to add an integration test in the server to show that... and what I got is a 401 :-P But I am not sure what I did wrong, I will need to take a deeper look ;-)

**Comment by skjnldsv** [comment]
> Regarding the fixes themselves the integration tests in the server pass, although some in Talk fail. One of those tests fail because, before, when you tried to update a received share (that you had access to) but the permissions did not allow you to do that you got a 403, but now you get a 404. I think that the previous value was the right one, as the share is found but you are forbidden to do the operation. However, I tried to add an integration test in the server to show that... and what I got is a 401 :-P But I am not sure what I did wrong, I will need to take a deeper look ;-)

I'm fine with that as well yes :)
I originally migrated the not found to forbidden for the cases I sthought were appropriate, but figured I could do it it another pr, so I reverted.
I guess I reverted too much and changed a case that was a forbidden before?
Let me know if you want me to do something else!

Thanks for the rebase :wink:

File: .drone.yml [hunk 2]
```diff
@@ -1193,19 +1193,19 @@ trigger:
 
 ---
 kind: pipeline
-name: integration-sharing-v1
+name: integration-checksums-v1
```
Comment (rullzer) [inline .drone.yml]: The changes in this file look suspicious....

Comment (danxuliu) [reply .drone.yml]: Let me quote myself ;-)

> reordered the pipelines to match the movement of files (`git diff --histogram` shows it better than the default diff algorithm)

**Comment by danxuliu** [comment]
I am working on this, please do not review yet.

File: .drone.yml (hunk 2 above)
Comment (rullzer) [reply .drone.yml]: O now I'm supposed to read your comments.
Next you want me to actually look at the code changes! It is madness I tell you! MADNESS

**Comment by danxuliu** [comment]
I have reordered the commits, split the code changes in different commits because I am too dumb to understand a single commit with all the changes, fixed the code to keep the old behaviour (404 status when trying to edit or delete a share that the user does not have access to) yet return 403 when the user has access to the file but is not allowed to edit or delete it, fixed deleting a group or room share as the owner/sharer, made `canAccessShare` and `canDeleteShare` independent from `canEditShare`, made some minor code style changes to adjust to code the rest of the file, fixed the unit tests, and added integration tests that broke with the previous code behaviour.

Despite that I left most comments and code from @skjnldsv ;-)

Only one integration test fails now in Talk, which needs to be adjusted (I will take care of that once this pull request is merged) as it fails because now a user with access to a file is able to get any of its shares.

Ready for review!

**Comment by skjnldsv** [comment]
Thanks a lot @danxuliu !!
Sorry you had to do such work, I thought I did a pretty good job with the unit and integration tests :(

**Review by skjnldsv — COMMENTED** [review]
:+1:

**Comment by rullzer** [comment]
Master is no 18 development. If this needs to go into 17 please follow the normal backport procedures after merging.

**Comment by skjnldsv** [comment]
Bump :bellhop_bell:

**Review by rullzer — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]
:elephant:

**Comment by skjnldsv** [comment]
Weird 
```
modified: apps/oauth2/js/oauth2.js
12 | modified: apps/oauth2/js/oauth2.js.map
13
```

**Comment by skjnldsv** [comment]
three acceptance failing, but when running locally:

```
15 scenarios (15 passed)
245 steps (245 passed)
2m53.09s (5.73Mb)

11 scenarios (11 passed)
91 steps (91 passed)
1m47.76s (5.73Mb)

12 scenarios (12 passed)
225 steps (225 passed)
2m57.98s (5.71Mb)
```

**Comment by skjnldsv** [comment]
> Weird
> 
> ```
> modified: apps/oauth2/js/oauth2.js
> 12 | modified: apps/oauth2/js/oauth2.js.map
> 13
> ```

So, even locally or the bot cannot find changes.
@rullzer @ChristophWurst if you compile this branch locally, do you find uncommited files?

**Comment by skjnldsv** [comment]
Okay, the acceptances issues are on master, not on this pr.
Still this weird issue on the oauth

**Pull request merged**